src/frontend/src/optimizer/plan_node/batch_sort_agg.rs

```diff
@@ -0,0 +1,148 @@
+        let ctx = logical.base.ctx.clone();
+        let input = logical.input();
+        let input_dist = input.distribution();
+        let dist = match input_dist {
+            Distribution::HashShard(_) => logical
+                .i2o_col_mapping()
+                .rewrite_provided_distribution(input_dist),
+            d => d.clone(),
+        };
+        let base =
+            PlanBase::new_batch(ctx, logical.schema().clone(), dist, logical.order().clone());
+        let input_ord = input.order();
+        let required_order = Order {
+            field_order: input_ord
+                .field_order
+                .iter()
+                .filter(|field_ord| {
+                    logical
+                        .group_key()
+                        .iter()
+                        .any(|g_k| *g_k == field_ord.index)
+                })
+                .cloned()
+                .collect(),
+        };
+
+        assert_eq!(required_order.field_order.len(), logical.group_key().len());
+
+        BatchSortAgg {
+            base,
+            logical,
+            required_order,
+        }
+    }
+
+    pub fn agg_calls(&self) -> &[PlanAggCall] {
+        self.logical.agg_calls()
+    }
+
+    pub fn group_key(&self) -> &[usize] {
+        self.logical.group_key()
+    }
+}
+
+impl fmt::Display for BatchSortAgg {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.logical.fmt_with_name(f, "BatchSortAgg")
+    }
+}
+
+impl PlanTreeNodeUnary for BatchSortAgg {
+    fn input(&self) -> PlanRef {
+        self.logical.input()
+    }
+
+    fn clone_with_input(&self, input: PlanRef) -> Self {
+        Self::new(self.logical.clone_with_input(input))
+    }
+}
+impl_plan_tree_node_for_unary! { BatchSortAgg }
+
+impl ToDistributedBatch for BatchSortAgg {
+    fn to_distributed(&self) -> Result<PlanRef> {
+        self.to_distributed_with_required(
+            &self.required_order,
+            &RequiredDist::shard_by_key(self.input().schema().len(), self.group_key()),
+        )
+    }
+
+    fn to_distributed_with_required(
+        &self,
+        required_order: &Order,
+        required_dist: &RequiredDist,
+    ) -> Result<PlanRef> {
+        let new_input = self
+            .input()
+            .to_distributed_with_required(required_order, required_dist)?;
+        Ok(self.clone_with_input(new_input).into())
+    }
+}
+
+impl ToBatchProst for BatchSortAgg {
+    fn to_batch_prost_body(&self) -> NodeBody {
+        NodeBody::SortAgg(SortAggNode {
+            agg_calls: self
+                .agg_calls()
+                .iter()
+                .map(PlanAggCall::to_protobuf)
+                .collect(),
+            group_key: self
+                .group_key()
+                .iter()
+                .clone()
+                .map(|index| ExprNode {
+                    expr_type: *index as i32,
+                    return_type: None,
+                    rex_node: None,
+                })
```

Comment:
we should make an `InputRef` expression here.

Reply:
`ExprImpl::InputRef(Box::new(InputRef::new())))` and `Expr::to_expr_proto`

Reply:
Thanks, for the hint!